pkg/ccl/kvccl/kvtenantccl/upgradeccl/tenant_upgrade_test.go

```diff
@@ -30,12 +30,321 @@ import (
+			Settings:          settings,
+			Knobs: base.TestingKnobs{
+				Server: &server.TestingKnobs{
+					DisableAutomaticVersionUpgrade: make(chan struct{}),
+					BinaryVersionOverride:          clusterversion.TestingBinaryMinSupportedVersion,
+				},
+			},
+		},
+	})
+	defer tc.Stopper().Stop(ctx)
+
+	expectedInitialTenantVersion := clusterversion.TestingBinaryMinSupportedVersion
+	expectedFinalTenantVersion := clusterversion.TestingBinaryMinSupportedVersion
+	connectToTenant := func(t *testing.T, addr string) (_ *gosql.DB, cleanup func()) {
+		pgURL, cleanupPGUrl := sqlutils.PGUrl(t, addr, "Tenant", url.User(username.RootUser))
+		tenantDB, err := gosql.Open("postgres", pgURL.String())
+		require.NoError(t, err)
+		return tenantDB, func() {
+			tenantDB.Close()
+			cleanupPGUrl()
+		}
+	}
+	firstServerSettings := cluster.MakeTestingClusterSettingsWithVersions(
+		clusterversion.TestingBinaryVersion,
+		clusterversion.TestingBinaryMinSupportedVersion,
+		false, // initializeVersion
+	)
+	secondServerSettings := cluster.MakeTestingClusterSettingsWithVersions(
+		clusterversion.TestingBinaryMinSupportedVersion,
+		clusterversion.TestingBinaryMinSupportedVersion,
+		false, // initializeVersion
+	)
+	require.NoError(t, clusterversion.Initialize(ctx,
+		expectedInitialTenantVersion, &firstServerSettings.SV))
+	require.NoError(t, clusterversion.Initialize(ctx,
+		expectedInitialTenantVersion, &secondServerSettings.SV))
+	mkTenant := func(t *testing.T, id uint64, clusterSettings *cluster.Settings) (tenantDB *gosql.DB, cleanup func()) {
+		tenantArgs := base.TestTenantArgs{
+			TenantID:     roachpb.MustMakeTenantID(id),
+			TestingKnobs: base.TestingKnobs{},
+			Settings:     clusterSettings,
+		}
+		tenant, err := tc.Server(0).StartTenant(ctx, tenantArgs)
+		require.NoError(t, err)
+		tenantDB, cleanup = connectToTenant(t, tenant.SQLAddr())
+		return tenantDB, cleanup
+	}
+
+	// Create a tenant and its first SQL server before upgrading anything and verify its version.
+	const tenantID = 10
+	tenantDB, firstServerCleanup := mkTenant(t, tenantID, firstServerSettings)
+	tenantRunner := sqlutils.MakeSQLRunner(tenantDB)
+
+	// Create a second SQL server with a lower binary version.
+	_, secondServerCleanup := mkTenant(t, tenantID, secondServerSettings)
+
+	// Ensure that the tenant works.
+	tenantRunner.CheckQueryResults(t, "SHOW CLUSTER SETTING version",
+		[][]string{{expectedInitialTenantVersion.String()}})
+	tenantRunner.Exec(t, "CREATE TABLE t (i INT PRIMARY KEY)")
+	tenantRunner.Exec(t, "INSERT INTO t VALUES (1), (2)")
+
+	// Upgrade the host cluster.
+	sqlutils.MakeSQLRunner(tc.ServerConn(0)).Exec(t,
+		"SET CLUSTER SETTING version = $1",
+		clusterversion.TestingBinaryVersion.String())
+
+	// Ensure that the tenant still works.
+	tenantRunner.CheckQueryResults(t, "SELECT * FROM t", [][]string{{"1"}, {"2"}})
+	// Give the tenant some time and ensure that it won't auto finalize
+	time.Sleep(time.Second * 30)
```

Comment:
Rather than this sleep, can we:

- Add a testing knob or package level testing function to speed up the auto-upgrade loop
- Add a testing knob that gets called in the auto-upgrade loop and perhaps gets passed the ultimate upgrade decisions

This would give us a way to then wait for the next call of the loop and make assertions about it.

Reply:
This is nice 👍 tests look much better now. Updated the tests to depend on a testing knob instead of actually waiting for the upgrade and also changed uses of `StartTestCluster` to `StartServerOnly`.